qf_lib/plotting/charts/waterfall_chart.py

```diff
@@ -15,61 +15,70 @@
 
 import numpy as np
 
-from plotting.decorators.coordinate import DataCoordinate
-from plotting.decorators.text_decorator import TextDecorator
+from qf_lib.plotting.decorators.coordinate import DataCoordinate
+from qf_lib.plotting.decorators.data_element_decorator import DataElementDecorator
+from qf_lib.plotting.decorators.text_decorator import TextDecorator
 from qf_lib.containers.series.qf_series import QFSeries
 from qf_lib.plotting.charts.chart import Chart
 
 
 class WaterfallChart(Chart):
-    def __init__(self, data: QFSeries, title: Optional[str] = None, percentage: Optional[bool] = False):
+    def __init__(self, percentage: Optional[bool] = False):
         super().__init__()
         self.total_value = None
         self.cumulative_sum = None
-        self.data = data
-        self.assert_is_qfseries(data)
-        self.title = title
         self.percentage = percentage
 
     def plot(self, figsize: Tuple[float, float] = None) -> None:
         self._setup_axes_if_necessary(figsize)
-        self.axes.set_xlim(0, self.data.size)
-
-        self.cumulative_sum = np.cumsum(self.data.values)
-        for index, value in enumerate(self.data.items()):
-            self._plot_waterfall(index, value)
+        self._configure_axis()
+        self._add_text()
+        self._apply_decorators()
 
-        # Set x-axis label using 'category' column
+    def _configure_axis(self):
+        data_element_decorators = self.get_data_element_decorators()
+        indices = [i for index in [d.data.index for d in data_element_decorators] for i in index]
+        self.axes.set_xlim(0, len(indices))
         self.axes.tick_params(axis='both', which='major', labelsize=10)
-        self.axes.set_xticks(range(len(self.data.index) + 2))
-        self.axes.set_xticklabels(['', *self.data.index, ''])
-        self.axes.set_title(self.title, y=1.04)
+        self.axes.set_xticks(range(len(indices) + 2))
+        self.axes.set_xticklabels(['', *indices, ''])
 
-        self._apply_decorators()
+    def _add_text(self):
+        data_element_decorators = self.get_data_element_decorators()
+        self.cumulative_sum = np.cumsum(np.concatenate([d.data.values for d in data_element_decorators]))
+        index = 0
+        for data_element in data_element_decorators:
```

Comment:
Let's use enumerate if we need index

Reply:
In this case we need to track the index over two for loops so the enumerate will not work